File: server/src/routes/forwards.ts
import type { FastifyInstance } from 'fastify';
import { z } from 'zod';
import type { ConnectionsResponse } from '@muxus/shared';
import type { AppContext } from '../app.js';
import { sendError } from '../util/errors.js';

const forwardRequestSchema = z.object({
  connId: z.string().min(1),
  type: z.enum(['local', 'remote', 'dynamic']),
  bindPort: z.number().int().min(1).max(65535),
  targetHost: z.string().min(1).optional(),
  targetPort: z.number().int().min(1).max(65535).optional(),
  tunnelId: z.string().min(1).optional(),
});

export function registerForwardRoutes(app: FastifyInstance, ctx: AppContext): void {
  app.get('/api/forwards', (req) => {
    const { connId } = req.query as { connId?: string };
    return { forwards: ctx.forwards.list(connId) };
  });

  app.post('/api/forwards', async (req, reply) => {
    try {
      const parsed = forwardRequestSchema.safeParse(req.body);
      if (!parsed.success) {
        return await reply.code(400).send({ message: parsed.error.issues[0]?.message ?? 'invalid forward request' });
      }
      return await ctx.forwards.start(parsed.data);
    } catch (err) {
      return sendError(reply, err);
    }
  });

  app.delete('/api/forwards/:id', (req) => {
    const { id } = req.params as { id: string };
    ctx.forwards.stop(id);
    return { ok: true };
  });

  // Adopt a running forward into a saved tunnel ("save as tunnel" without
  // restarting the listener).
  app.patch('/api/forwards/:id', async (req, reply) => {
    const { id } = req.params as { id: string };
    const parsed = z.object({ tunnelId: z.string().min(1) }).safeParse(req.body);
    if (!parsed.success) {
      return reply.code(400).send({ message: 'tunnelId is required' });
    }
    const info = ctx.forwards.assignTunnel(id, parsed.data.tunnelId);
    if (!info) return reply.code(404).send({ message: 'forward not found' });
    return info;
  });

  // Live SSH transports — lets the forwarding panel reuse an existing
  // connection for a tunnel instead of dialing a second one.
  app.get('/api/connections', (): ConnectionsResponse => ({ connections: ctx.connections.list() }));
}

File: app/internal/server/routes/forwards.go
package routes

import (
	"encoding/json"
	"net/http"

	"github.com/go-chi/chi/v5"

	"github.com/FloSch62/muxus/app/internal/api"
	"github.com/FloSch62/muxus/app/internal/forwards"
	"github.com/FloSch62/muxus/app/internal/server"
)

func writeForwardError(w http.ResponseWriter, err error) {
	if problem, ok := err.(*forwards.HTTPError); ok {
		server.WriteJSON(w, problem.Status, api.ErrorBody{Message: problem.Message})
		return
	}
	writeInternalError(w, err)
}

func validForwardRequest(request api.ForwardRequest) bool {
	if request.ConnID == "" || request.BindPort < 1 || request.BindPort > 65535 {
		return false
	}
	if request.Type != "local" && request.Type != "remote" && request.Type != "dynamic" {
		return false
	}
	if request.Type != "dynamic" {
		return request.TargetHost != "" && request.TargetPort >= 1 && request.TargetPort <= 65535
	}
	return true
}

func RegisterForwardRoutes(r chi.Router, ctx *server.Context) {
	r.Get("/api/forwards", func(w http.ResponseWriter, req *http.Request) {
		server.WriteJSON(w, http.StatusOK, map[string]any{
			"forwards": ctx.Forwards.List(req.URL.Query().Get("connId")),
		})
	})
	r.Post("/api/forwards", func(w http.ResponseWriter, req *http.Request) {
		body, ok := readJSONBody(w, req, defaultBodyLimit)
		if !ok {
			return
		}
		var input api.ForwardRequest
		if err := json.Unmarshal(body, &input); err != nil || !validForwardRequest(input) {
			writeBadRequest(w, "invalid forward request")
			return
		}
		info, err := ctx.Forwards.Start(input, "manual")
		if err != nil {
			writeForwardError(w, err)
			return
		}
		server.WriteJSON(w, http.StatusOK, info)
	})
	r.Delete("/api/forwards/{id}", func(w http.ResponseWriter, req *http.Request) {
		ctx.Forwards.Stop(chi.URLParam(req, "id"))
		server.WriteJSON(w, http.StatusOK, map[string]bool{"ok": true})
	})
	r.Patch("/api/forwards/{id}", func(w http.ResponseWriter, req *http.Request) {
		body, ok := readJSONBody(w, req, defaultBodyLimit)
		if !ok {
			return
		}
		var input struct {
			TunnelID string `json:"tunnelId"`
		}
		if json.Unmarshal(body, &input) != nil || input.TunnelID == "" {
			writeBadRequest(w, "tunnelId is required")
			return
		}
		info := ctx.Forwards.AssignTunnel(chi.URLParam(req, "id"), input.TunnelID)
		if info == nil {
			writeNotFound(w, "forward not found")
			return
		}
		server.WriteJSON(w, http.StatusOK, info)
	})
	r.Get("/api/connections", func(w http.ResponseWriter, _ *http.Request) {
		server.WriteJSON(w, http.StatusOK, api.ConnectionsResponse{
			Connections: ctx.Connections.List(),
		})
	})
}
